content: Add draft of the Source track. (#1037)
slsa-framework/slsa: docs/spec/v1.1/source-requirements.md

# SLSA Source Track

## Objective

> The SLSA Source Track mitigates [Threat A ("Submit unauthorized change")](/spec/v1.0/threats#a-submit-unauthorized-change), scoped to a code repository and the organization that owns that repository. Concretely: an attacker must compromise the accounts of two organization members to publish code in a Source Level 3-conformant repository, and the evidence of those unauthorized changes cannot be destroyed without further attacks.

Reviewer: (The "Threat A" url references a deleted ref.)

It seems to me that the source track (similar to the build track) should start with generating provenance at the lower levels, then move to implementation of specific policies at higher levels. 
The revision of a consumed repository is a kind of artifact like any other software dependency and as such should come with all its paperwork. The "build process" of a repository revision is the pull request, and the result of a completed pull request should be a new revision accompanied by provenance claims. 

This example appears to be covering the requirements and benefits of a specific policy, similar to the ["Build Level 1" dedicated infrastructure requirement](https://slsa.dev/spec/v1.0/levels#build-l1). 
> Concretely: an attacker must compromise the accounts of two organization members to publish code in a Source Level 3-conformant repository, and the evidence of those unauthorized changes cannot be destroyed without further attacks.

At a minimum, consumers of a repository revision will want to know the following provenance claims: 
1. Who contributed these changes (according to the SCP)? 
2. When did they do so (according to the SCP)? 
3. What policy compliance claims are associated with this revision (according to the SCP)? 

Based on this, it seems to me that a focused adaptation of the excellent ["build track" summary](https://slsa.dev/spec/v1.0/levels#build-track) would work well here. EG:

The SLSA **SOURCE** track describes increasing levels of trustworthiness and completeness in a **repository revision's** provenance. Provenance describes what **Source Control Provider produced the revision**, what process was used, and who the contributors were. The lowest level only requires the provenance to exist, while higher levels provide increasing protection against tampering of the **version control system, the provenance, or the revision**.

The primary purpose of the **source** track is to enable [verification](https://slsa.dev/spec/v1.0/verifying-artifacts) that the **revision followed the expected process**. Consumers have some way of knowing what the expected provenance should look like for a given package and then compare each **source revision's** actual provenance to those expectations. Doing so prevents several classes of [supply chain threats](https://slsa.dev/spec/v1.0/threats).

Reviewer 2: Something like that seems reasonable to me.

The one question I have is, you describe the provenance as indicating who the contributors were for a given revision, is that meant to exclude some attestation that describes the state of the repo as a whole?

I think we also need something that describes the state of a the repository as a whole over its lifetime so that consumers don't _need_ to process each specific revision.

Perhaps the intent is to think of 'provenance' broadly here where it might include a couple of different attestation types to achieve those goals?  In which case we'd just need to tweak the language some to make that more clear?

On a related note: do we like the term 'provenance' here?  Will that wind up being confused with https://slsa.dev/provenance

Reviewer 2: I'm also realizing that the build track specifies how the information about the build is communicated (provenance) but the source track currently doesn't.  That seems like a gap that should be filled (which would hopefully help address some of these points?)

Reviewer: > is that meant to exclude some attestation that describes the state of the repo as a whole

Yes, there are (typically) many commits in a repo that are invalid, illegal, incomplete or malicious. When you checkout a repo revision for a build, you'll mostly want to know about the version you got. Info about other branches, other revisions, other commits, other points in time is probably not relevant to the choice about whether or not _this_ revision is good to use. 

> something that describes the state of the repository as a whole over its lifetime
Can you talk a little more about how you'd use this kind of data? I guess I'm thinking mostly of ci/cd scenarios: EG, if I fetched a revision of a repo with sha abc123. I'm not sure it would be very relevant to me to know things about the whole repo, like whether the main ref is protected, or when the rules were last changed for the dev branch, etc. Mostly I'll want to know which rules were processed to allow the current revision I have to exist.

Reviewer 2: > Can you talk a little more about how you'd use this kind of data? I guess I'm thinking mostly of ci/cd scenarios: EG, if I fetched a revision of a repo with sha abc123. I'm not sure it would be very relevant to me to know things about the whole repo, like whether the main ref is protected, or when the rules were last changed for the dev branch, etc. Mostly I'll want to know which rules were processed to allow the current revision I have to exist.

Yes exactly.  And that's what the original (3 years old 😨 ) [Source Control Attestation proposal](https://docs.google.com/document/d/10B-8R3NIGvcPuoIhqb-p18oOe78m6SiiukfE7NyvnYQ/edit?usp=sharing) was trying to get at.

Which isn't to say that it should exclude some ability to reason about any particular commit, they're just separate problems.

As I understand it (and I haven't been paying strict attention to this effort) the latest thinking was that it's useful to have _both_ capabilities.  E.g. 1. some ability for a consumer to determine if the current revision of source is 'good' and 2. some ability for motivated systems or people to determine _why_ 1 is true.

I noticed one of the changes to this PR actually removed the notional attestation format from this PR, so it looks like resolution of that problem is being deferred?

Reviewer 2: FWIW that doc is shared with https://groups.google.com/g/slsa-discussion.  Everyone should be able to join that group to read it.  My work org doesn't let me share docs with the entire public :(.

## Changes from v0.1

-   **Scope** The Source track is now scoped to Revisions rather than builds.
Why?: To facilitate verification without anchoring it to a build.

-   **Model** Added a model, definitions, and the concept of verification.
Why?: SLSA does not yet have a model for version control systems, and we need such a model to be able to discuss them.

## Outstanding TODOs

- [] Flesh out the definition and bounds of 'identity', and why they're required.
- [] Refine requirements/guidance for trusted robots.
- [] Either identify the unique value of L1 or merge it with L2.

## Source model

The Source track is scoped to a single project that is controlled by some organization. That organization determines what Source level should apply to the project and administers technical controls to enforce that level.

| Term | Description
| --- | ---
| Source | An identifiable set of text and binary files and associated metadata usually used as input for the build system (see SLSA Build Track).
| Organization | A collection of people who collectively create the Source. Examples of organizations include an open-source projects, a company, or a team within a company.

> | Change | A set of modifications to one or more source files and associated metadata. Change metadata MUST include any information required to situate the change in relation to othe| Version Control System | Software for tracking and managing changes to source. Git and Subversion are examples of version control systems.

Reviewer: ```suggestion
| Change | A set of modifications to one or more source files and associated metadata. Change metadata MUST include any information required to situate the change in relation to other changes (e.g. parent revision).
| Version Control System | Software for tracking and managing changes to source. Git and Subversion are examples of version control systems.
```

> | Revision | The canonical source at a given point in time as identified by the version control system. As an example, you can identify a git revision by its tree hash.

Reviewer 3: I agree git does not have a verifiable timestamp but SCP should have one ? (so maybe the example is wrong). Time is an important point in case you detect an attack such as XZ and need to go back to a "safe" point.

> | Revision | The canonical source at a given point in time as identified by the version control system. As an example, you can identify a git revision by its tree hash.

Reviewer: ```suggestion
| Revision | A specific state of a repository as identified by the version control system. As an example, you can identify revision of a git repository by its tree id. 
```

> | Revision | The canonical source at a given point in time as identified by the version control system. As an example, you can identify a git revision by its tree hash.

Reviewer: @Nikokrock it's true that the SCP should have a timestamp for when a revision was uploaded, but 1) not all SCPs provide that feature, and 2) (to me) that information seems separate from what defines the revision itself. I think I can upload the same revision of a repository to multiple SCPs. They would all have different "receive" timestamps but should be considered "the same code" imo.

> r changes (e.g. parent revision).

Reviewer: ```suggestion
```

### Source Roles

| Role | Description
| --- | ---
| Administrator | A human who can perform privileged operations on one or more projects. Privileged actions include, but are not limited to, modifying the change history and modifying project- or organization-wide security policies.
| Trusted person | A human who is authorized by the organization to propose and approve changes to the source.
| Trusted robot | Automation with an authentic identity that is authorized by the organization to propose and/or approve changes to the source.
| Untrusted person | A human who has limited access to the project. They MAY be able to read the source. They MAY be able to propose or review changes to the source. They MAY NOT approve changes to the source or perform any privileged actions on the project.
| Proposer | The role that proposes a particular change to the source.
| Reviewer | The role that reviews a particular proposed change to the source.
| Approver | The role that approves a particular change to the source.

> | Merger | The role that applies a change to the source. This person may be the proposer or a different trusted person, depending on the version control platform.

Reviewer: There is an actor who "clicked the merge button", but this may or may not be the identity of the actor that authenticated to push the changes. 

For example, I may click the merge button to add a PR to a GitHub Merge Queue, but the Merge Queue's service principal is the identity who eventually pushes the ref update of the main branch.
This is because merge-queue-like systems contain the contents of multiple contributors and change sets.

Reviewer 2: I think the intent here is whomever pushed the merge button (they eventually cause the Merge Queue to update the main branch).

Do you have any suggestions for alternate language?

Reviewer 2: Maybe "The _identity of the person_ that applies a change to the source. This person may be the proposer or a different trusted person, depending on the version control platform." ?

Reviewer: It seems to me that protected refs should not be writable by actors other than the SCP service principal (side note: this is unfortunately not the case for gh today).
If a protected ref _is_ written by something other than the SCP SP, that's generally a big red flag. 
 
I'd propose "Merger: the human or Trusted Robot that applies a change to the source."

Git digression: 
For most of these security levels, you'd want this to exclusively be the SCP. If a random human was the last pusher to a release ref, you'd probably want to go look through audit logs to see what happened, why did they bypass pr rules, etc.

One downside of this approach is that it makes your push history look very robotic.
We handle mq merges with special ux treatment to help call out that the mq ref was full of mergeable prs.

Reviewer 2: So I think we might be talking about different things, and maybe we want to capture both?

1. The service principle (am I using this right?) of the entity (user or robot) that actually authorized the changes to the underlying VCS.
2. The service principle of the entity that _triggered_ the change in the first place (which might operate in practice by putting some work on a queue for the trusted robot you mention)

Num 1 is helpful if we care about which accounts have broad scope to tamper with source.

Num 2 is helpful if we want the accounts involved in the review of the change itself.

So in the case you describe (1) would be the SCP SP, (2) would be the account of the person that caused the SCP SP to make that change.

## Source Platform Requirements

> The version control system MUST provide at least:

Reviewer: This section may be mixing concerns between the VCS and the source control platform (SCP). 

Git commit metadata is forgeable from the perspective of the SCP. 
The SCP typically provides an identity / authz layer which is separate from the git content it manages, and it maintains its own timestamps for when activities happened from its perspective. EG: a commit may claim it was authored last year, but all the SCP knows for sure is that it was uploaded last week. 

Authentic contributions will typically be built in terms of the SCP's identity and timestamp models, not the VCS's.

Reviewer 2: Is the best fix to remove the distinction between the two in this section and state that it's all the responsibility of the SCP who _may_ implement these features via the VCS as appropriate?

    -   The immutable reference to the new revision.
    -   The identities of the proposer, reviewers (if any), and merger (if different to the proposer).
    -   Timestamps of change submission. If a change is reviewed, then the change history MUST also include timestamps for any reviews.
    -   The change description/justification.
    -   The content of the change.
    -   The parent revisions.

Most popular version control systems meet these requirement, such as git, Subversion, Mercurial, and Perforce.

The source control platform MUST provide at least:

-   An account system or some other means of identifying persons.
-   A mechanism for modifying the canonical source through a **revision process**.

The source control platform SHOULD additionally provide:

-   A mechanism for assigning roles and/or permissions to identities.
-   A mechanism for including code review in the revision process.
-   Two-factor authentication for the account system (L2+ only).
-   Audit logs for sensitive actions, such as modifying security controls.

## Levels

### Level 1: Version controlled

Summary: The project source is stored and managed through a modern version control system.

> Intended for: Organizations that are unwilling or unable to host their source on a source control platform. If possible, skip to Level 2.

Reviewer: > host their source on a source control platform

Would this include self-hosted versions of SCPs?

Reviewer 2: IMO this should only apply to folks who are using 'bare' VCS and that self-hosted SCPs should be eligible for higher levels with the caveat that the trust story will be murkier.  E.g. much easier for a consumer to 'trust' the hosted GitHub/GitLab/etc... solution vs the one I've spun up myself.  But it shouldn't be impossible.

Seem fair?

Reviewer 4: I think self-hosted should be permitted, thinking specifically of projects like freedesktop.org and gnome.org. 

As we're focused on preventing malicious changes through a single account compromise I think an important aspect of the SCP is that the developer account/identity can't make arbitrary/non-observable changes the SCP in order to introduce changes which otherwise wouldn't be allowed. Ideally this would be a achieved through separate ops/admin account, or better yet team, to manage the SCP.

Requirements:

**[Version controlled]** Every change to the source is tracked in a version control system that meets the requirements listed in [Source Platform Requirements](#source-platform-requirements).

Benefits: Version control solves software development challenges from ranging change attribution to effective collaboration. It is a software development best practice with more benefits than we can discuss here.

### Level 2: Verified history

Summary: The source code and its change history metadata are retained and authenticated to allow trustworthy auditing and analysis of the source code.

Intended for: Organizations that are unwilling or unable to incorporate code review into their software development practices.

Requirements:
**[Strong authentication]** User accounts that can modify the source or the project's configuration must use multi-factor authentication or its equivalent.

**[Verified timestamps]** Each entry in the change history must contain at least one timestamp that is determined by the source control platform and cannot be modified by clients. It MUST be clear in the change history which timestamps are determined by the source control platform.

> **[Retained history]** The change history MUST be preserved as long as the source is hosted on the source control system. The source MAY migrate to another source control system, but the organization MUST retain the change history if possible. It MUST NOT be possible for persons to delete or modify the change history, even with multi-party approval, except by trusted platform admins following an established deletion policy.

Reviewer: verified timestamps are provided by the SCP, so when I migrate the source to a new SCP, I think this means that the org must keep the original repo around to keep the "true" timestamps. 
Or else, we may be saying that SCPs must provide a way to export the verified timestamps somehow. We should consider making that clearer here, and I'm not sure which way is better! I'm not sure SCPs would import timestamps like that.

Reviewer 2: Does this come back to the question above of how important the actual time is? (I can see that going either way).

Assuming it _is_ important we would seem to need some ability to track the original times when moving between SCPs?  Should we be evaluating some option for preserving this data that doesn't rely on the SCP itself?  E.g. if we had a per commit attestation a Time Stamp Authority could provide a verified timestamp that could easily migrate to a new SCP.

On a related note, in "MUST retain the change history if possible" I find "MUST" and "if possible" to be contradictory?  Do we resolve this problem by just downgrading "MUST" to "SHOULD"?

Reviewer 2: So I see your point, it's not strictly contradictory; but I think it is confusing?

In this case it's certainly _possible_ for an SCP to implement some feature that imports the time in a secure fashion and we'll be asking the SCPs to implement other things to implement this spec.  So why bother with the "if possible" qualifier?  Should we be more specific on what cases don't need to retain this information?  Perhaps "importing code from an SCP that does not support or does not have change history".

Reviewer 5: Upon further reflection I think you're right. 
Even if "MUST... if possible" can be interpreted as stronger than SHOULD the reality is that SHOULD already carry the intent we want and there is no real value in getting into the details of the cases where this might not be done.
For the sake of simplicity let's just change it to a SHOULD and drop the "if possible".

Benefits: Attributes changes in the version history to specific actors and timestamps, which allows for post-auditing, incident response, and deterrence for bad actors. Multi-factor authentication makes account compromise more difficult, further ensuring the integrity of change attribution.

### Level 3: Changes are authorized

Summary: All changes to the source are approved by two trusted persons prior to submission.

Intended for: Enterprise projects and mature open source projects.

Requirements:

> **[Code review]** All changes to the source are approved by two trusted persons prior to submission. User accounts that can perform code reviews MUST use two-factor authentication or its equivalent.

Reviewer: Should a robot approval count or must it always be human?

Reviewer 3: From my point, it is not always a human (some sources might not be easily read by a human, for example). Or in case of merges, depending on the way you handle your merge processes, a robot might be able to review if the change can be approved or not.

Reviewer 4: We've talked in the past about how we might modify review constraints in the face of robots, I believe it was deemed out of scope for this first PR. Previous discussion includes:

- https://github.com/slsa-framework/slsa/issues/196
- https://github.com/slsa-framework/slsa/issues/956
- https://github.com/slsa-framework/slsa/issues/287

I know some organisations allow robots to automatically merge dependency updates if the change doesn't introduce a breakage. I think that makes sense so long as there are additional controls (i.e., the WIP dependency track in #961) on what changes the robot can introduce.

The following combinations of trusted persons are acceptable:

-   Proposer and reviewer are two different trusted persons.
-   Two different reviewers are trusted persons.

The code review system must meet the following requirements:

-   **[Informed review]** The reviewer is able and encouraged to make an informed decision about what they're approving. The reviewer MUST be presented with a full, meaningful content diff between the proposed revision and the previously reviewed revision. For example, it is not sufficient to just indicate that a file changed without showing the contents.

> -   **[Context-specific approvals]** Approvals are for a specific context, such as a repo + target branch + revision in git. Moving fully reviewed content from one context to another still requires review, except for well-understood automatic processes. For example, you do not need to review each change to cut a release branch, but you do need review when backporting changes from the main branch to an existing release branch.

Reviewer: > you do not need to review each change to cut a release branch

This seems really tricky. 
If I create a new release ref that points to the tip of main, I must ensure that the tip of main complies with all requirements that govern the release ref. 
It's normal for the requirements to be different between those contexts. 

> Moving fully reviewed content from one context to another still requires review, except for well-understood automatic processes

This makes sense to me. I'm not sure the example expresses this though.

-   **[Atomic change sets]** Changes are recorded in the change history as a single revision that consists of the net delta between the proposed revision and the parent revision. In the case of a nonlinear version control system, where a revision can have more than one parent, the diff must be against the "first common parent" between the parents. In other words, when a feature branch is merged back into the main branch, only the merge itself is in scope.

> Trusted robots MAY be exempted from the code review process. It is RECOMMENDED that trusted robots so exempted be run only software built at Build L3+ from sources that meet Source L3.

Reviewer: ```suggestion
Trusted robots MAY be exempted from the code review process. It is RECOMMENDED that an exempted trusted robot artifact be built at Build L3+ from sources that meet Source L3.
```
